PDEP-1 Revision (Decision Making) (#53576)
pandas-dev/pandas: web/pandas/pdeps/0001-purpose-and-guidelines.md

The file is a markdown file, you can use `web/pdeps/0001.md` as a reference
for the expected format.

The initial status of a PDEP will be `Status: Draft`. This will be changed to
`Status: Under discussion` by the author(s), when they are ready to proceed with the decision
making process.

#### PDEP Discussion Timeline

A PDEP discussion will remain open for up to 60 days. This period aims to enable participation
from volunteers, who might not always be available to respond quickly, as well as provide ample
time to make changes based on suggestions and considerations offered by the participants.
Similarly, the following voting period will remain open for 15 days.

To enable and encourage discussions on PDEPs, we follow a notification schedule. At each of the
following steps, the pandas team, and the pandas dev mailing list are notified via GitHub and
E-mail:
- Once a PDEP is ready for discussion.
- After 30 discussion days, with 30 days remaining for discussion.
- After 45 discussion days, with 15 days remaining for discussion.

> - In case 15 days passed without any new comments, the authors may close the discussion period

Reviewer: Can we add ``close the discussion period preemptively`` here?

> - In case 15 days passed without any new comments, the authors may close the discussion period

Author: ```suggestion
- In case 15 days passed without any new comments, the authors may close the discussion period preemptively, 
```

  and open the voting period.
- Once the voting period starts, after 60 days or in case of an earlier vote, with 15 days
  remaining for voting.
- After 10 voting days, with 5 days remaining for voting.

#### Casting Votes

> As the voting period starts, a VOTE issue is created which links to the PDEP discussion pull request.

Reviewer 2: If the VOTE issues are public (with anyone allowed to post) it might be worth commenting that VOTE issues will state their intention to delete any vote that is not valid, i.e. if it is posted by a non-voting member?

Reviewer 3: I think that's a good addition

> Each voting member may cast a vote by adding one of the following comments:

Reviewer 2: Apologies if I havce missed this but it does appear that a "voting member" is explicitly defined?

The removed section seems to suggest that a voting member is anyone in the "core development team". Does pandas have any segregation between "active" and "inactive" core development members and does this have any bearing on  their right to vote?

Is there any scope to specifically invite certain informed (non-core dev) parties to vote on issues? I would probably consider this an abuse of power and instead invite to participate in the discussion to influence members.

If votes are cast by core dev members does this have implications for inviting future members into the core dev team? i.e to not push for adding members of a group of similar, or friendly like minded individuals, perhaps from the same institution etc, that can build up substantial iunfluence?

Reviewer 4: > Right now, it is the core team, but it is likely we will not use that in the future.

IMO we should also require a discussion to change this, hence my comment to define voting member in this PDEP now.

- +1: approve.

> - 0: abstain.

Reviewer 3: That's correct.

> - 0: abstain.

Reviewer 2: Could we add this to the doc to make it explicit, and also could we add to the tallies so that the votes are more transparent, i.e. have Approved-Abstained-Disapproved-NoResponse.

It may be relevant for audit at some point in the future to see who / how many people were eligible to vote.

  - Reason: A one sentence reason is required.
- -1: disapprove
  - Reason: A one sentence reason is required.

> A disapprove vote requires prior participation in the PDEP discussion issue.

Reviewer 2: Would it not be sufficient for a voting member to follow a discussion led by knowledgeable parties and simply provide a vote that read: "-1: disapprove. I agree with XXX's analysis and concerns and disagree that YYY's argument and the PDEP are suitable".

Rather than provide rules that restrict voting in certain ways (which might be difficult to police anyway), would it not be sufficient to design quorum and majority rules that aim to progress PDEPs that have garnered 'sufficient' support (whatever that may end up being), above objections.

Once the voting period ends, any voter may tally the votes in a comment, using the format: x-y-z,
where x stands for the total of approving, y of abstaining, and z of disapproving votes cast.

#### Quorum and Majority

For a PDEP vote to result in accepting the proposal, a quorum is required. All votes (including
abstentions) are counted towards the quorum. The quorum is computed as the lower of these two
values:

> - 11 voting members.

Reviewer 5: IDK if this has been asked before, but can you vote for your own PDEP?

(Getting a quorum might be more challenging for PDEPs with e.g. 2/3 authors)

Author: I assumed the authors would vote - yes. Happy to make that explicit.

- 50% of voting members.

> Given a quorum, a majority of 75% of the non-abstaining votes is required as well, i.e. 75% of

Reviewer: Scikit learn seems to require a simple majority

> Given a quorum, a majority of 75% of the non-abstaining votes is required as well, i.e. 75% of

Author: We discussed this in our governance meeting today, and at least within the group today, we all found the 75% appropriate. 

The 75% came from a consensus seeking decision-making approach, see [this wiki](https://en.wikipedia.org/wiki/Consensus_decision-making#Decision_rules).
The idea was that we don't to have a decision which had many negative votes. 

Personally, I also like the "unanimity minus three" approach - so that once 4 people are against something, the group shouldn't accept it. But I think it needs to appeal to everyone, so it would be great to get more opinions.

> Given a quorum, a majority of 75% of the non-abstaining votes is required as well, i.e. 75% of

Reviewer 6: We also though it could be helpful to make the 75% number more tangible with a concrete example: if 12 people vote, 4 negative votes would block it; if 16 people vote, 5 negative votes would block it.

(and so this current proposal is less strict than a "unanimity minus three" rule)

the approving and disapproving votes must be in favor.

Thus, abstaining votes count towards a quorum, but not towards a majority. A voting member might
choose to abstain when they have participated in the discussion, have some objections to the
proposal, but do not wish to stop the proposal from moving forward, nor indicate their full
support.

> 

Reviewer 2: Should there be an explicit statement discussing what happens if a quorum is not reached, i.e. if not enough voting members participate in the vote?
Does that mean the proposal is explicitly rejected and cannot be resubmitted?
Or does it mean that the PDEP may undergo a period of hiberation followed by a new period of discussion and a new vote is permitted after at least (4?) months of the previous vote?

Reviewer: Why shouldn't it be possible to resubmit pdeps if circumstances change? How would decide if it's the same pdep or something different if some parameters are changed? I'd rather not worry about this now. Most PDEPs will be submitted by core-devs, and I think there isn't anyone in the group who wants to spam us with proposals

Reviewer 3: > Should there be an explicit statement discussing what happens if a quorum is not reached, i.e. if not enough voting members participate in the vote?

We haven't discussed this possibility.

Author: We discussed this in our governance meeting today. We propose that not reaching quorum is equivalent to rejecting because effectively the PDEP was not accepted in time. The authors may decide to resubmit any rejected PDEP, as long as they add a link to the previous discussion in the new PDEP proposal. As before, we expect PDEP authors to use their own judgement when they decide to resubmit a PDEP proposal (or withdraw one), as to the ongoing relevance and likelihood of reaching quorum and/or acceptance for the resubmission. No hibernation period is explicitly required.
I'll add a change to reflect this soon.

The file is a markdown file, you can use `web/pdeps/0001.md` as a reference
for the expected format.

The initial status of a PDEP will be `Status: Draft`. This will be changed to
`Status: Under discussion` by the author(s), when they are ready to proceed with the decision
making process.

#### PDEP Discussion Timeline

A PDEP discussion will remain open for up to 60 days. This period aims to enable participation
from volunteers, who might not always be available to respond quickly, as well as provide ample
time to make changes based on suggestions and considerations offered by the participants.
Similarly, the following voting period will remain open for 15 days.

To enable and encourage discussions on PDEPs, we follow a notification schedule. At each of the
following steps, the pandas team, and the pandas dev mailing list are notified via GitHub and
E-mail:
- Once a PDEP is ready for discussion.
- After 30 discussion days, with 30 days remaining for discussion.
- After 45 discussion days, with 15 days remaining for discussion.

> - In case 15 days passed without any new comments, the authors may close the discussion period

Author: ```suggestion
- In case 15 days passed without any new unaddressed comments, the authors may close the discussion period
```

> - In case 15 days passed without any new unaddressed comments, the authors may close the discussion period

Reviewer: I have a question about this: Is it required to give notice that the voting will start in x days or can you just say voting starts now? This is a bit confusing if that happens very early because discussion did not gain traction.

Reviewer 3: The concept is that you can say "voting starts now"

Reviewer: Can we require at least a discussion period of 3 weeks? I was a bit surprised that this was already ready. Granted, it was my fault to not object to the 75% majority earlier (or follow up on my comment above), but I expected that I'd have more time.

My reasoning here is: Get a earliest moment in time when the voting procedure could possibly start

Reviewer 3: Well, the way it is written is that if there is no discussion for 15 days, you could call for a vote.  So in theory, if there was no discussion for 15 days right after a PDEP was created, the author could call for a vote, but I think that is highly unlikely.

Or we can say that there is a minimum discussion period of 3 weeks, as you suggest.

Reviewer: I'd like the 3 weeks, it seems like a low requirement as you said and would make things clearer. I previously assumed that it would require at least 45 days of discussion (that was on me though).

Reviewer 6: > Can we require at least a discussion period of 3 weeks? I was a bit surprised that this was already ready. 

FWIW, this PR wasn't necessarily "ready", it's just that my question if there were more comments triggered people to start approving .. 

But +1 on requiring at least 3 weeks (which already had passed in this case, though)

> - In case 15 days passed without any new unaddressed comments, the authors may close the discussion period

Author: ```suggestion
- After 45 days, in case 15 days passed without any new unaddressed comments, the authors may close the discussion period
```

Reviewer 4: Since there is a requirement to participate in order to downvote, then we could have a situation where a discussion has not had any objections and proceeding to a vote would be pointless since only those in support could vote.

This is maybe not a bad thing as it puts the onus on members to get involved in the discussion early.

Extending the minimum discussion period is a potential solution to give sufficient opportunity to participate, but it maybe that we should have a notice period for the vote to begin. "voting starts now" is a bit final.

Reviewer 4: to avoid this then maybe we should just remove the requirement to participate in the discussion in order to be able to downvote

Reviewer 6: > Extending the minimum discussion period is a potential solution to give sufficient opportunity to participate, but it maybe that we should have a notice period for the vote to begin. "voting starts now" is a bit final.

It's a good point that when a discussion has been quiet, being able to just "start the voting" with an early vote, sounds a bit abrupt (exactly because the discussion went quiet, people might have forgotten about commenting on the PDEP).

I updated the text around this slightly to just state in general that we expect a notification 15 days before a vote would start. And then this heads up for a nearing vote can either be after 30 days for an early vote (so the end result is the same as Noa's suggestion above that the actual cote can happen after 45 days) or later.

Reviewer 4: Thanks @jorisvandenbossche 

We use the term "unaddressed comment" in the proposal. I changed my review here to "requires changes" recently because of this.

Now, IIRC in the discussion about PDEP-8 (i've not checked the timelines thoroughly and it maybe that I should use a hypothetical case instead) that we had potentially progressed to the vote with unaddressed comments.

Any ideas to cover this scenario or whether we need to at this stage?

> To enable and encourage discussions on PDEPs, we follow a notification schedule. At each of the
> following steps, the pandas team, and the pandas dev mailing list are notified via GitHub and
> E-mail:
> - Once a PDEP is ready for discussion.
> - After 30 discussion days, with 30 days remaining for discussion.
> - After 45 discussion days, with 15 days remaining for discussion.
> - In case 15 days passed without any new unaddressed comments, the authors may close the discussion period

Reviewer 3: ```suggestion
To enable and encourage discussions on PDEPs, we follow a notification schedule. At each of the
following steps, the pandas team, and the pandas dev mailing list are notified via GitHub and
E-mail, with an indication that a voting period will commence afterr15 days of the notification date if there is no further discussion:
- Once a PDEP is ready for discussion.
- After 30 discussion days, with 30 days remaining for discussion.
- After 45 discussion days, with 15 days remaining for discussion.
```

Reviewer 6: @Dr-Irv I pushed a different commit with an attempt to address the same thing, see https://github.com/pandas-dev/pandas/pull/53576/commits/88a7db99fef424bc6614d4ba536a6b3587ff7140. Can you take a look at that?

Reviewer 3: @jorisvandenbossche  I looked at just that commit and made a comment on that text.

  preemptively, and open the voting period.
- Once the voting period starts, after 60 days or in case of an earlier vote, with 15 days
  remaining for voting.
- After 10 voting days, with 5 days remaining for voting.

#### Casting Votes

As the voting period starts, a VOTE issue is created which links to the PDEP discussion pull request.

> Each voting member, including the author(s) may cast a vote by adding one of the following comments:

Author: ```suggestion
Each voting member, including author(s) with voting rights, may cast a vote by adding one of the following comments:
```

- +1: approve.

> - 0: abstain.

Reviewer 6: Do we want to add a small explanation of how we understand "abstaining" or what it could mean? 
Something like "Not fully convinced, but don't want to block it", or better worded (although there might be many reasons to vote +0, like "I haven't closely followed, but trust the majority")

  - Reason: A one sentence reason is required.
- -1: disapprove
  - Reason: A one sentence reason is required.
A disapprove vote requires prior participation in the PDEP discussion issue.

Comments made on the public VOTE issue by non-voting members will be deleted.

Once the voting period ends, any voter may tally the votes in a comment, using the format: w-x-y-z,

> where w stands for the total of approving, x of abstaining, z of disapproving votes cast, and z

Reviewer 3: ```suggestion
where w stands for the total of approving, x of abstaining, y of disapproving votes cast, and z
```

- 3 August 2022: Initial version ([GH-47938][47938])
- 15 February 2023: Version 2 ([GH-51417][51417]) clarifies the scope of PDEPs and adds examples

> - DD MM YY: Version 3(link) defines a structured decision making process for PDEPs

Reviewer 3: ```suggestion
- 09 June 2023: Version 3([GH-53576][53576]) defines a structured decision making process for PDEPs
```
